### DISCUSSION TITLE
Proper way to reuse Redis connections?

### DISCUSSION THREAD

**Discussion by immasandwich** [vercel/vercel#5134] [Help]
Hey everyone, 

Love working with Vercel. However, we have a really high traffic website right now that hits a our API endpoints and is thoroughly trashing our maximum concurrent Redis connections on LambdaStore.

```ts
export default async (req: NextApiRequest, res: NextApiResponse) => {
  const client: redis.RedisClient = redis.createClient({
    port: 31066,
    host: "my.lambda.store",
    password: "some-password",
  });

  // ...do something
});
```

I've seen some solutions that suggest to pull the client _outside_ of the handler, so it becomes part of a shared Execution Context (at least, that's how it is described in AWS Lambda), as so:

```ts
const client: redis.RedisClient = redis.createClient({
  port: 31066,
  host: "my.lambda.store",
  password: "some-password",
});

export default async (req: NextApiRequest, res: NextApiResponse) => {
  // ...do something
});
```

I hit my cloud function with Artillery (tiny load test: 20 requests per second, 10 seconds), and yet I still see the number of connections still increasing on LambdaStore dashboard.

Clearly, we're not understanding the shared context as its being used for a shared MySQL connection here: https://vercel.com/guides/deploying-next-and-mysql-with-vercel (although this uses `serverless-mysql`, it is in effect theorizing the same strategy... keep a pool of instances that can be referenced across executions).

Can anyone explain how shared contexts work in Vercel's lambda functions, and how we can manage the number of connections to continue to scale?

**Comment by pip8786** [comment]
@kshahani did you find a way to do this? I'm interested in what a solution would be.

**Comment by chrsep** [comment]
I encountered a similar issue when i have to connect to redis on my side project. What seems to help for me was to try to find and kill any zombie connections left after each redis query, inspired by how serverless-mysql's [`.end()`](https://github.com/jeremydaly/serverless-mysql/blob/abb64ceaa9cfb8470caacfae4fe04ff6f71f6d14/index.js#L127-L160) works, which you need to call after running your queries.

If i understand it correctly, lambda functions runs inside containers, shared context are only shared by lambda functions running on the same containers. What causes the connection overload was that for some reason a lot of separate containers were created.

For now, I ended up creating a `killZombies()` function that I call at the end of my queries, it basiclly kill the connections with the highest idle time when the number of connected client is above a certain threshold.
```typescript
const getClients = async () => {
  const result: string = await c.client("list")

  return result
    .split("\n") // split by line
    .filter((text) => text) // removes empty value
    .map((client) => {
      // turn array of "key=value" string into "{key: value}" object
      const attributes = client.split(" ")
      const clientObject: any = {}
      attributes.forEach((attribute) => {
        const [key, value] = attribute.split("=")
        clientObject[key] = value
      })
      return clientObject
    })
    .sort((a, b) => b.idle - a.idle)
}

const killZombies = async () => {
  const clients = await getClients()

  const maxClientCount = (MAX_CONNECTIONS * CONNECTION_THRESHOLD) / 100
  if (clients.length > maxClientCount) {
    const clientsToKill = clients.length - maxClientCount

    const killCommands: Promise<void>[] = []
    for (let i = 0; i < clientsToKill; i += 1) {
      killCommands.push(c.client("kill", ["id", clients[i].id]))
    }

    const result = await Promise.all(killCommands)
  }
}
```
This sometimes can kill connections that are still being used tho. Those clients would then need to reconnect. In my case this was handled automatically by ioredis. There might be better ways to reduce zombies.

**Reply by pip8786** [reply]
I wonder, if you expect to be running in a serverless environment like Vercel, if you should just kill your connection after you've rendered the page. Is the expectation that your next command might run on a different container and therefore it's not useful to keep that initial connection open? I guess since it runs across many containers, you can't really share a connection between all the different function calls.

**Reply by chrsep** [reply]
Based on the logs that I got from my project, it seems that lambda does reuse existing containers when it can, so closing connections directly after the function finish running might not be ideal. But sometimes a lot of containers can pop up. I'm not sure yet what all the causes are.

Though, one thing that i know can trigger an onslaught of new containers is if a bunch of concurrent requests were to come before any lambda containers become ready to be re-used, each of those request would trigger the creation of its own containers. The discussion on this [stackoverflow answer](https://stackoverflow.com/a/45782990/6656573) might explained it better.

**Comment by fs-c** [comment]
I'm not absolutely sure if this is relevant for this specific issue, but it helped me with a very similar one. I use MongoDB and until recently created a connection for every execution of every serverless function. This, predictably, demolished the weak little VPS I used to host the DB.

My partial solution:

```js
// test/store.js
let cache;

const id = require('crypto').randomBytes(2).toString('hex');

exports.get = () => {
    if (cache) {
        console.log(id, 'returning from cache')

        return cache;
    }

    console.log(id, 'calculating');

    return cache = 5;
};


// test/a.js
const store = require('./store');

module.exports = (req, res) => {
    res.json({ value:  store.get() });
};

// test/b.js
const store = require('./store');

module.exports = (req, res) => {
    res.json({ value:  store.get() });
};
```

This is an excerpt from the logs, I was alternating between hitting `/api/test/a` three times and hitting `/api/test/b` three times:

```
event - build page: /api/test/a
wait  - compiling...
event - compiled successfully
677e calculating                         # first hit on a, calculated for the first time
677e returning from cache                # 2nd hit, cache
677e returning from cache                # 3rd hit, cache
event - build page: /api/test/b
wait  - compiling...
event - compiled successfully
b054 calculating                         # first hit on b, calculated for the second time in total
b054 returning from cache                # 2nd hit, cache
b054 returning from cache                # 3rd hit, cache
b054 returning from cache                # 4th hit on a, cache
b054 returning from cache                # 5th hit, cache
b054 returning from cache                # 6th hit, cache
b054 returning from cache                # 4th hit on b, cache
b054 returning from cache                # ...
b054 returning from cache
```

You can see that it's only calculated if the endpoint hasn't been called before, and that all functions share the latest store. Assuming you don't clean up dead connections, the maximum number of open connections using this method is therefore the number of serverless functions using the store.
